osu.Game/Tests/Visual/ModSandboxTestScene.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using JetBrains.Annotations;
+using osu.Game.Beatmaps;
+using osu.Game.Replays;
+using osu.Game.Rulesets;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Scoring;
+using osu.Game.Screens.Play;
+
+namespace osu.Game.Tests.Visual
+{
+    public abstract class ModSandboxTestScene : PlayerTestScene
+    {
+        protected sealed override bool HasCustomSteps => true;
+
+        public override IReadOnlyList<Type> RequiredTypes => new[]
+        {
+            typeof(ModSandboxTestScene)
+        };
+
+        protected ModSandboxTestScene(Ruleset ruleset)
+            : base(ruleset)
+        {
+        }
+
+        private ModTestCaseData currentTest;
+
+        protected void CreateModTest(ModTestCaseData testCaseData) => CreateTest(() =>
+        {
+            AddStep("set test data", () => currentTest = testCaseData);
+        });
+
+        public override void TearDownSteps()
+        {
+            AddUntilStep("test passed", () => currentTest?.PassCondition?.Invoke() ?? true);
+            base.TearDownSteps();
+        }
+
+        protected sealed override IBeatmap CreateBeatmap(RulesetInfo ruleset) => currentTest?.Beatmap ?? base.CreateBeatmap(ruleset);
+
+        protected sealed override Player CreatePlayer(Ruleset ruleset)
+        {
+            SelectedMods.Value = SelectedMods.Value.Append(currentTest.Mod).ToArray();
+
+            var score = currentTest.Autoplay
+                ? ruleset.GetAutoplayMod().CreateReplayScore(Beatmap.Value.GetPlayableBeatmap(ruleset.RulesetInfo, SelectedMods.Value))
+                : new Score { Replay = new Replay() };
```

Comment:
What issues does using the actual autoplay mod cause?

Reply:
Autoplay mod does not fails, for instance. It also doesn't allow restarting upon failure.

Both of those make SD and PF impossible to test.